File: app-builder/frontend/src/shared/runtime/pageAccess.ts
import type { Page, PageAccess, PageAccessMode } from '../schema/types'

export const DEFAULT_PAGE_ACCESS: PageAccess = { mode: 'public' }

export type PageAccessResolution = {
  pageId: string | null
  redirected: boolean
  unavailable: boolean
}

export function normalizePageAccess(value: unknown): PageAccess {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return { ...DEFAULT_PAGE_ACCESS }
  const access = value as Record<string, unknown>
  const mode = normalizePageAccessMode(access.mode)
  const redirectPageId = typeof access.redirectPageId === 'string'
    ? access.redirectPageId.trim()
    : ''

  return {
    mode,
    ...(redirectPageId ? { redirectPageId } : {}),
  }
}

export function isPageAccessible(
  page: Pick<Page, 'access'>,
  signedIn: boolean,
): boolean {
  const mode = normalizePageAccess(page.access).mode
  if (mode === 'signedIn') return signedIn
  if (mode === 'signedOut') return !signedIn
  return true
}

export function resolvePageAccess(
  pages: Array<Pick<Page, 'id' | 'access'>>,
  requestedPageId: string | null | undefined,
  signedIn: boolean,
): PageAccessResolution {
  if (pages.length === 0) {
    return { pageId: null, redirected: false, unavailable: true }
  }

  const pagesById = new Map(pages.map((page) => [page.id, page]))
  const requestedPage = requestedPageId ? pagesById.get(requestedPageId) : undefined
  let currentPage: Pick<Page, 'id' | 'access'> | undefined = requestedPage ?? pages[0]
  const visited = new Set<string>()

  while (currentPage && !visited.has(currentPage.id)) {
    visited.add(currentPage.id)
    if (isPageAccessible(currentPage, signedIn)) {
      return {
        pageId: currentPage.id,
        redirected: currentPage.id !== requestedPageId,
        unavailable: false,
      }
    }

    const redirectPageId: string | undefined = normalizePageAccess(currentPage.access).redirectPageId
    currentPage = redirectPageId ? pagesById.get(redirectPageId) : undefined
  }

  const fallback = pages.find((page) => isPageAccessible(page, signedIn))
  return {
    pageId: fallback?.id ?? null,
    redirected: Boolean(fallback && fallback.id !== requestedPageId),
    unavailable: !fallback,
  }
}

function normalizePageAccessMode(value: unknown): PageAccessMode {
  if (value === 'signedIn' || value === 'signedOut') return value
  return 'public'
}

File: app-builder/native-preview/Android/app/src/main/java/com/apptura/nativepreview/navigation/PageAccess.kt
package com.apptura.nativepreview.navigation

import com.apptura.nativepreview.models.Page

internal fun isPageAccessible(page: Page, signedIn: Boolean): Boolean {
    return when (normalizePageAccessMode(page.access?.mode)) {
        "signedIn" -> signedIn
        "signedOut" -> !signedIn
        else -> true
    }
}

internal fun resolveAccessiblePageIndex(
    pages: List<Page>,
    requestedPageId: String?,
    signedIn: Boolean,
): Int? {
    if (pages.isEmpty()) return null

    val pagesById = pages.associateBy { it.id }
    var currentPage = requestedPageId?.let(pagesById::get) ?: pages.first()
    val visited = mutableSetOf<String>()

    while (visited.add(currentPage.id)) {
        if (isPageAccessible(currentPage, signedIn)) {
            return pages.indexOfFirst { it.id == currentPage.id }.takeIf { it >= 0 }
        }

        val redirectPageId = currentPage.access?.redirectPageId?.trim().orEmpty()
        currentPage = pagesById[redirectPageId] ?: break
    }

    return pages.indexOfFirst { isPageAccessible(it, signedIn) }.takeIf { it >= 0 }
}

private fun normalizePageAccessMode(mode: String?): String {
    return when (mode) {
        "signedIn", "signedOut" -> mode
        else -> "public"
    }
}
